src/main/java/uk/nhs/tis/trainee/notifications/service/HistoryService.java

```diff
@@ -166,6 +167,21 @@ public List<HistoryDto> findAllForTrainee(String traineeId) {
         .toList();
   }
 
+  /**
+   * Find all sent historic notifications for the given Trainee.
+   *
+   * @param traineeId The ID of the trainee to get notifications for.
+   * @return The found notifications, empty if none found.
+   */
+  public List<HistoryDto> findAllSentForTrainee(String traineeId) {
+    List<History> history = repository.findAllByRecipient_IdOrderBySentAtDesc(traineeId);
+
+    return history.stream()
+        .filter(h -> Instant.now().isAfter(h.sentAt()))
+        .map(this::toDto)
+        .toList();
```

Comment:
As the stream is ordered by the SentAt date with notification history which is not ready to be displayed at the back. So I use `.takeWhile()` instead to pick from front of the stream.

Reply:
I don't think this is right.

The results of the repository call returns the history entities order by `sentAt` in **descending** order e.g.
  1. `2024-12-12`
  2. `2023-06-06`
  3. `2022-01-01`

so using `takeWhile()` will drop everything if there are any future values for `sentAt`.

The unit test appears to return mocked values in the wrong order, but if you add an integration test based on `HistoryServiceIntegrationTest.shouldSortFoundNotificationsBySentAtWhenMultipleFound()` you'll see it fail with your current logic if you change the assertions to the below

```java
List<HistoryDto> foundHistory = service.findAllSentForTrainee(TRAINEE_ID);

assertThat("Unexpected history count.", foundHistory.size(), is(2));

HistoryDto history2 = foundHistory.get(0);
assertThat("Unexpected history sent at.", history2.sentAt(), is(now));

HistoryDto history3 = foundHistory.get(1);
assertThat("Unexpected history sent at.", history3.sentAt(), is(before));
```

If you change the stream processing to the below it will pass

```java
Instant now = Instant.now();
return history.stream()
    .dropWhile(h -> h.sentAt().isAfter(now))
    .map(this::toDto)
    .toList();
```